x-pack/libbeat/common/nomad/watcher.go

```diff
@@ -0,0 +1,222 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package nomad
+
+import (
+	"fmt"
+	"time"
+
+	api "github.com/hashicorp/nomad/api"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// Max back off time for retries
+const maxBackoff = 30 * time.Second
+
+// Watcher watches Nomad task events
+type Watcher interface {
+	// Start watching Nomad API for new events
+	Start() error
+
+	// Stop watching Nomad API for new events
+	Stop()
+
+	// AddEventHandler add event handlers for handling specific events
+	AddEventHandler(ResourceEventHandlerFuncs)
+}
+
+type watcher struct {
+	client    *api.Client
+	options   WatchOptions
+	logger    *logp.Logger
+	nodeID    string
+	lastFetch time.Time
+	handler   ResourceEventHandlerFuncs
+	waitIndex uint64
+	done      chan struct{}
+}
+
+// WatchOptions controls watch behaviors
+type WatchOptions struct {
+	// SyncTimeout is a timeout for tasks
+	SyncTimeout time.Duration
+	// Node is used for filtering events
+	Node string
+	// Namespace is used for filtering events on specified namespacesx,
+	Namespace string
+	// RefreshInterval is the time interval that the Nomad API will be queried
+	RefreshInterval time.Duration
+	// AllowStale allows any Nomad server (non-leader) to service
+	// a read. This allows for lower latency and higher throughput
+	AllowStale bool
+	// InitialWaitIndex specify the initial WaitIndex to send in the first request
+	// to the Nomad API
+	InitialWaitIndex uint64
+}
+
+// NewWatcher initializes the watcher client to provide a events handler for
+// resource from the cluster (filtered to the given node)
+func NewWatcher(client *api.Client, options WatchOptions) (Watcher, error) {
+	w := &watcher{
+		client:    client,
+		options:   options,
+		logger:    logp.NewLogger("nomad"),
+		waitIndex: options.InitialWaitIndex,
+		lastFetch: time.Now(),
+		done:      make(chan struct{}),
+	}
+
+	queryOpts := &api.QueryOptions{
+		WaitTime:   w.options.SyncTimeout,
+		AllowStale: w.options.AllowStale,
+		WaitIndex:  uint64(1),
+	}
+
+	if w.nodeID == "" {
```

Comment:
We initially considered the option of _streaming_ the logs instead of running filebeat in each host, which meant that it could potentially run on a host without an agent, but given the current implementation, this sounds better.

Reply:
Yes, I agree with keeping an option to configure this for scenarios like the one you mention where the agent is not running or is not reachable. But taking the id from the host in the agent, if available, looks like a nice default.